Makefile

```diff
@@ -90,6 +110,80 @@ test-silent:
 
+define dockerbuild
+	docker run \
+		--net none \
+		--user "$(USERID)" \
+		--volume "$(PWD)/out:/out" \
+		--volume "$(PWD):/go/src/github.com/aws/amazon-ecs-agent" \
+		--rm \
+		$(BUILDER_IMAGE) \
+		$(1)
+endef
+
+define win-cgo-dockerbuild
+	docker run --net=none \
+		--user "$(USERID)" \
+		--volume "$(PWD)/out:/out" \
+		--volume "$(PWD):/go/src/github.com/aws/amazon-ecs-agent" \
+		--env "GOOS=windows" \
+		--env "CGO_ENABLED=1" \
+		--env "CC=x86_64-w64-mingw32-gcc" \
+		--rm \
+		$(BUILDER_IMAGE) \
+		$(1)
+endef
+
+# TODO: use `go list -f` to target the test files more directly
+ALL_GO_FILES = $(shell find . -name "*.go" -print | tr "\n" " ")
+GO_INTEG_TEST = go test -race -tags integration -c -o
+out/test-artifacts/linux-engine-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_INTEG_TEST) $@ ./agent/engine)
+
+out/test-artifacts/linux-app-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_INTEG_TEST) $@ ./agent/app)
+
+out/test-artifacts/linux-stats-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_INTEG_TEST) $@ ./agent/stats)
+
+out/test-artifacts/windows-engine-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_INTEG_TEST) $@ ./agent/engine)
+
+out/test-artifacts/windows-app-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_INTEG_TEST) $@ ./agent/app)
+
+out/test-artifacts/windows-stats-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_INTEG_TEST) $@ ./agent/stats)
+
+GO_FUNCTIONAL_TEST = go test -tags functional -c -o
+out/test-artifacts/linux-simple-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_FUNCTIONAL_TEST) $@ ./agent/functional_tests/tests/generated/simpletests_unix)
+
+out/test-artifacts/linux-handwritten-tests: out $(ALL_GO_FILES)
+	$(call dockerbuild,$(GO_FUNCTIONAL_TEST) $@ ./agent/functional_tests/tests)
+
+out/test-artifacts/windows-simple-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_FUNCTIONAL_TEST) $@ ./agent/functional_tests/tests/generated/simpletests_windows)
+
+out/test-artifacts/windows-handwritten-tests.exe: out $(ALL_GO_FILES)
+	$(call win-cgo-dockerbuild,$(GO_FUNCTIONAL_TEST) $@ ./agent/functional_tests/tests)
+
+.PHONY: test-artifacts-windows test-artifacts-linux test-artifacts
+
+WINDOWS_ARTIFACTS_TARGETS := out/test-artifacts/windows-engine-tests.exe out/test-artifacts/windows-stats-tests.exe
+WINDOWS_ARTIFACTS_TARGETS += out/test-artifacts/windows-app-tests.exe out/test-artifacts/windows-simple-tests.exe
+WINDOWS_ARTIFACTS_TARGETS += out/test-artifacts/windows-handwritten-tests.exe
+
+LINUX_ARTIFACTS_TARGETS := out/test-artifacts/linux-engine-tests out/test-artifacts/linux-stats-tests
+LINUX_ARTIFACTS_TARGETS += out/test-artifacts/linux-app-tests out/test-artifacts/linux-simple-tests
+LINUX_ARTIFACTS_TARGETS += out/test-artifacts/linux-handwritten-tests
+
+test-artifacts-windows: $(WINDOWS_ARTIFACTS_TARGETS)
+
+test-artifacts-linux: $(LINUX_ARTIFACTS_TARGETS)
```

Comment:
Running `make test-artifacts-linux` twice results in the builds happening twice. This is probably the case with `test-artifacts-windows` as well. It would be nice if this was not the case.

Reply:
The reason is that I didn't understand [order-only prerequisites](https://www.gnu.org/savannah-checkouts/gnu/make/manual/html_node/Prerequisite-Types.html). Because `out` is labeled phony, it gets executed every time, leading make to believe that test-artifacts-linux needs to be executed as well. 

I can solve this by marking `out` as order-only.

Reply:
Do you really need order-only prereqs, or can you use stamp files and non-PHONY targets?